## 1. Intro
* Premier League Player Analysis

## 2. Import Libraries

In [106]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

## 3. Data Loading

In [107]:
#epl_data = pd.read_csv('data/EPL_22_23.csv', encoding = 'unicode_escape')
epl_data = pd.read_csv('data/EPL_22_23.csv')

## 4. Data Overview

In [108]:
# Display the first few rows of the dataset
epl_data.head()

Rk            Player   Nation    Pos         Squad  Age  Born  MP  Starts  \
0   1  Brenden Aaronson   us USA  MF,FW  Leeds United   21  2000  36      28   
1   2     George Abbott  eng ENG     MF     Tottenham   16  2005   1       0   
2   3         Che Adams  sct SCO     FW   Southampton   26  1996  28      23   
3   4       Tyler Adams   us USA     MF  Leeds United   23  1999  24      24   
4   5  Tosin Adarabioyo  eng ENG     DF        Fulham   24  1997  25      23   

     Min   90s  Goals  Assists  G+A  G-PK  PK  PKatt  CrdY  CrdR  
0  2,372  26.4      1        3    4     1   0      0     2     0  
1      1   0.0      0        0    0     0   0      0     0     0  
2  1,992  22.1      5        3    8     5   0      0     1     0  
3  2,156  24.0      0        0    0     0   0      0     6     1  
4  2,085  23.2      1        0    1     1   0      0     4     0

In [109]:
# Check the dimensions of the dataset
epl_data.shape

(569, 19)

As can be seen, the dataset contains 569 rows and 19 columns.

#### Column Descriptions
Rk - Rank
Nation - Nationality of the player.
Pos - Position
	GK - Goalkeepers
	DF - Defenders
	MF - Midfielders
	FW - Forwards
	FB - Fullbacks
	LB - Left Backs
	RB - Right Backs
	CB - Center Backs
	DM - Defensive Midfielders
	CM - Central Midfielders
	LM - Left Midfielders
	RM - Right Midfielders
	WM - Wide Midfielders
	LW - Left Wingers
	RW - Right Wingers
	AM - Attacking Midfielders
Age -- Age at season start
MP -- Matches Played
Matches Played by the player or squad
Starts -- Game or games started by player
Min -- Minutes
90s -- 90s Played
Gls -- Goals
Ast -- Assists
G+A -- Goals + Assists
G-PK -- Non-Penalty Goals
PKatt -- Penalty Kicks Attempted
CrdY -- Yellow Cards
CrdR -- Red Cards

## 5. Data Cleaning and Preprocessing

In [110]:
# Display the column labels
epl_data.columns

Index(['Rk', 'Player', 'Nation', 'Pos', 'Squad', 'Age', 'Born', 'MP', 'Starts',
       'Min', '90s', 'Goals', 'Assists', 'G+A', 'G-PK', 'PK', 'PKatt', 'CrdY',
       'CrdR'],
      dtype='object')

In this analysis, we don't need all 19 columns, so we only need to select the columns we need.

In [111]:
epl_data = epl_data[['Player', 'Squad', 'Nation', 'Pos', 'Age', 'MP', 'Starts', 'Min', '90s','Goals', 'Assists', 'PK', 'PKatt', 'CrdY', 'CrdR']]

# add 2 columns using sum()
#epl_data['red'] = epl_data[['CrdY', '2CrdY']].sum(axis=1)

epl_data.head()


Player         Squad   Nation    Pos  Age  MP  Starts    Min  \
0  Brenden Aaronson  Leeds United   us USA  MF,FW   21  36      28  2,372   
1     George Abbott     Tottenham  eng ENG     MF   16   1       0      1   
2         Che Adams   Southampton  sct SCO     FW   26  28      23  1,992   
3       Tyler Adams  Leeds United   us USA     MF   23  24      24  2,156   
4  Tosin Adarabioyo        Fulham  eng ENG     DF   24  25      23  2,085   

    90s  Goals  Assists  PK  PKatt  CrdY  CrdR  
0  26.4      1        3   0      0     2     0  
1   0.0      0        0   0      0     0     0  
2  22.1      5        3   0      0     1     0  
3  24.0      0        0   0      0     6     1  
4  23.2      1        0   0      0     4     0

Now let’s check if there are any null values in the dataset. This can be simply done by using .isnull() method. is.null() returns True if a certain datapoint is Null and returns False otherwise. The number of null values can be calculated as seen below.

In [112]:
epl_data.isnull().sum()

Player     0
Squad      0
Nation     0
Pos        0
Age        0
MP         0
Starts     0
Min        0
90s        0
Goals      0
Assists    0
PK         0
PKatt      0
CrdY       0
CrdR       0
dtype: int64

Hooray! We can see there are no null values.

Let’s look at the overall data types using .info() function.

In [113]:
epl_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 15 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Player   569 non-null    object 
 1   Squad    569 non-null    object 
 2   Nation   569 non-null    object 
 3   Pos      569 non-null    object 
 4   Age      569 non-null    int64  
 5   MP       569 non-null    int64  
 6   Starts   569 non-null    int64  
 7   Min      569 non-null    object 
 8   90s      569 non-null    float64
 9   Goals    569 non-null    int64  
 10  Assists  569 non-null    int64  
 11  PK       569 non-null    int64  
 12  PKatt    569 non-null    int64  
 13  CrdY     569 non-null    int64  
 14  CrdR     569 non-null    int64  
dtypes: float64(1), int64(9), object(5)
memory usage: 66.8+ KB


The variable ‘Min’(minutes played) is an object type. We therefore convert this into integer type for future analysis by using .astype() method. Before using .astype(), we remove the comma in the values of ‘Min’.

In [114]:
epl_data['Min'] = epl_data['Min'].replace(',', '', regex = True)

In [115]:
epl_data.head()

Player         Squad   Nation    Pos  Age  MP  Starts   Min  \
0  Brenden Aaronson  Leeds United   us USA  MF,FW   21  36      28  2372   
1     George Abbott     Tottenham  eng ENG     MF   16   1       0     1   
2         Che Adams   Southampton  sct SCO     FW   26  28      23  1992   
3       Tyler Adams  Leeds United   us USA     MF   23  24      24  2156   
4  Tosin Adarabioyo        Fulham  eng ENG     DF   24  25      23  2085   

    90s  Goals  Assists  PK  PKatt  CrdY  CrdR  
0  26.4      1        3   0      0     2     0  
1   0.0      0        0   0      0     0     0  
2  22.1      5        3   0      0     1     0  
3  24.0      0        0   0      0     6     1  
4  23.2      1        0   0      0     4     0

If you look at the column ‘Nation’, you can see there are two abbreviation for each country. For example, for England, there are both ‘eng’ and ‘ENG’, and for Brazil, there are ‘br’ and ‘BRA’ for Brazil. Instead of having two abbreviations, we want to leave the capitalized abbreviation for each country.

The same for the column ‘Pos’. There are some players who are capable of playing multiple positions such as Bukayo Saka in the first row. We will select the first position only for convenient analysis.

In [116]:
epl_data['Nation'] = epl_data['Nation'].apply(lambda x: x.split(' ')[1])
epl_data['Pos'] = epl_data['Pos'].apply(lambda x: x.split(',')[0])
epl_data.head()

Player         Squad Nation Pos  Age  MP  Starts   Min   90s  \
0  Brenden Aaronson  Leeds United    USA  MF   21  36      28  2372  26.4   
1     George Abbott     Tottenham    ENG  MF   16   1       0     1   0.0   
2         Che Adams   Southampton    SCO  FW   26  28      23  1992  22.1   
3       Tyler Adams  Leeds United    USA  MF   23  24      24  2156  24.0   
4  Tosin Adarabioyo        Fulham    ENG  DF   24  25      23  2085  23.2   

   Goals  Assists  PK  PKatt  CrdY  CrdR  
0      1        3   0      0     2     0  
1      0        0   0      0     0     0  
2      5        3   0      0     1     0  
3      0        0   0      0     6     1  
4      1        0   0      0     4     0

Now since all the preprocessing is done, we move on to exploring the data.

## 6. Exploratory Data Analysis (EDA)

#### Where are the players from?

Let’s first have a look at the ‘Nation’ column. In python, value_counts() is a function that allows you to observe the counts of unique values. Using value_counts() with some visualization tools such as Seaborn and Matplotlib, we can plot a bar graph showing the distribution of nationality in the Premier League.

In [117]:
top_nations = epl_data['Nation'].value_counts()[:10]
df = pd.DataFrame({'Nation': top_nations.index, 'Frequency': top_nations.values})

# Create the bar chart
fig = px.bar(df, x='Nation', y='Frequency', title='Players Nationality', text_auto=True)

# Customize the layout to mimic ggplot style
fig.update_layout(
    xaxis_title="Nation",
    yaxis_title="Frequency",
    template="ggplot2"
)

# Show the plot
fig.show()

As quite expected, there were many English players in the EPL 2022–23 season, followed by Brazilian, Spanish and French players.

#### How old are the players?

Now let’s have a look at the distribution of player ages using sns.histplot().

In [118]:
#plt.figure(figsize = (8, 5))
#sns.histplot(epl_data['Age'], kde = True)
#plt.title('EPL players Age distribution')
#plt.show()

fig = px.histogram(epl_data.Age, x="Age", text_auto=True)
fig.update_layout(bargap=0.2)
fig.update_layout(title=dict(text="EPL players Age distribution"), bargap=0.2)
# Add X and Y labels
fig.update_xaxes(title_text="Age")
fig.update_yaxes(title_text="Number of Players")

fig.show()

#### Who is the best goal scorer?

Now let’s try to see who scored the most goals in 2021–22 season in the EPL.

In [119]:
epl_data[epl_data['Goals'] == epl_data['Goals'].max()]

Player            Squad Nation Pos  Age  MP  Starts   Min   90s  \
217  Erling Haaland  Manchester City    NOR  FW   22  35      33  2769  30.8   

     Goals  Assists  PK  PKatt  CrdY  CrdR  
217     36        8   7      7     5     0

Erling Haaland from Manchester City scored 36 goals in the 2022-2023 season.

In [120]:
cards_df = epl_data.groupby(['Squad']).agg({'CrdY':'sum','CrdR':'sum'}).sort_values(by = 'CrdY', ascending = False)
cards_df.head()

CrdY  CrdR
Squad                      
Leeds United       89     3
Wolves             85     6
Nott'ham Forest    84     0
Crystal Palace     82     3
Chelsea            81     3

Leeds United received the most yellow cards during the season with 89 yellow cards!

#### Top 5 Defenders Players that have the most Red Cards

In [121]:
DF_players = epl_data[epl_data['Pos'].str.contains("DF")]
DF_players_red = epl_data.nlargest(5, 'CrdR')[['Player', 'CrdR', 'CrdY']]

fig = px.bar(DF_players_red, x="Player", y=["CrdR", "CrdY"],
             color_discrete_map={
                 "CrdR": "#D83F31",
                 "CrdY": "#E9B824"},
             text_auto=True

             )
fig.update_layout(title_text='Top 5 Defenders Players with the most Red Cards',
                  title_x=0.5, title_font=dict(size=20), bargap=0.2)

# Add X and Y labels
fig.update_xaxes(title_text="Player")
fig.update_yaxes(title_text="Count")

fig.show()

#### Top 5 Players that have the scored the most Goals

In [122]:
fig_bar = px.bar(data_frame=epl_data.nlargest(5, 'Goals')[['Player', 'Goals']],
                 x='Player', y='Goals', color='Goals', text='Goals')
fig_bar.update_layout(title_text='Top 5 Players Goal Scorers',  # Main title for the project
                      title_x=0.5, title_font=dict(size=20))  # Location and the font size of the main title
fig_bar.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig_bar.show()

**According to graph above, the player who scored the most Goals this season is Erling Haaland with 36 goals. Harry Kane followed him with 30 goals and so on.**

#### Top 5 Players wih the most Assists

In [123]:
fig_bar = px.bar(data_frame=epl_data.nlargest(5, 'Assists')[['Player', 'Assists']],
                 x='Player', y='Assists', color='Assists', text='Assists')
fig_bar.update_layout(title_text='Top 5 Players with the most Goal Assists',  # Main title for the project
                      title_x=0.5, title_font=dict(size=20))  # Location and the font size of the main title
fig_bar.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig_bar.show()

**According to graph above, the player who had the most Goal Assists this season is Kevin De Bruyne with 16. Mohamed follows him with 12 assists**

#### Graph of Players who have the highest Penalty_Goals/Penalty_Attempted Ratio

In [124]:
PenaltyPerAttempted = pd.concat([epl_data['Player'], 1/(epl_data['PKatt']/epl_data['PK'])], axis=1)
PenaltyPerAttempted = PenaltyPerAttempted.replace([np.inf], np.nan).dropna(axis=0)
PenaltyPerAttempted['Percentage'] = PenaltyPerAttempted[0] * 100

fig = px.bar(
    data_frame=PenaltyPerAttempted,
    x='Player',
    y=0,
    color=0,
    hover_data={'Percentage': ':.2f%'},  # Add hover data with percentage format
    labels={'0': 'Ratio'}  # Optionally rename the y-axis label
)

fig.update_layout(
    title_text='Graph of Players who have the highest Penalty_Goals/Penalty_Attempted Ratio',
    title_x=0.5,
    title_font=dict(size=20),
    xaxis={'categoryorder': 'total descending'}
)

fig.update_yaxes(tickformat='.0%')  # Update y-axis to show values as percentages
fig.show()